### PR TITLE
[FLINK-5810] [flip6] Introduce a hardened slot manager

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR is based on #3310.

Harden the slot manager so that it better deals with lost and out of order messages
from the TaskManager. The basic idea is that the TaskManager are considered the ground
truth and the SlotManager tries to maintain a consistent view of what is reported to it
by the TaskManagers. This has the assumption that the TaskManagers regularly report their
slot status to the SlotManager piggy backed on the heartbeat signals to the ResourceManager (not yet implemented, though). That way it is possible to handle lost and out of order messages because the SlotManager will eventually converge on a consistent view of the actual slot allocation.

Additionally, the hardened SlotManager registers for idle TaskManagers and pending slot
requests a timeout. If the timeout expires, then the TaskManagers are released and the
slot requests are failed, respectively. This prevents resource leaks and wasteful resource allocation.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java [hunk 1]
```diff
@@ -21,519 +21,897 @@
...
+ * allocated slot is freed, then it tries to fulfill another pending slot request. Whenever there
+ * are not enough slots available the slot manager will notify the resource manager about it via
+ * {@link ResourceManagerActions#allocateResource(ResourceProfile)}.
+ *
+ * In order to free resources and avoid resource leaks, idling task managers (task managers whose
+ * slots are currently not used) and not fulfilled pending slot requests time out triggering their
+ * release and failure, respectively.
  */
-public abstract class SlotManager {
+public class SlotManager implements AutoCloseable {
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: Implementing `AutoCloseable` is a neat trick, but I find it almost a bit confusing. Would one ever use a SlotManager in a "try-with-resource" statement?

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java [hunk 2]
```diff
@@ -21,519 +21,897 @@
...
 		}
 
-		// try to fulfil the request with current free slots
-		final ResourceSlot slot = chooseSlotToUse(request, freeSlots);
-		if (slot != null) {
-			LOG.info("Assigning SlotID({}) to AllocationID({}), JobID:{}", slot.getSlotId(),
-				allocationId, request.getJobId());
+		pendingSlotRequests.clear();
+
+		HashSet<InstanceID> registeredTaskManagers = new HashSet<>(taskManagerRegistrations.keySet());
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: If the copy of the collection is just for concurrent iteration, I would make it an `ArrayList`. Builds faster and iterates faster.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java [hunk 3]
```diff
@@ -21,519 +21,897 @@
...
-				ResourceSlot slot = checkNotNull(getRegisteredSlot(slotId));
-				// treat this slot as empty and retry with a different request
-				handleFreeSlot(slot);
+	public boolean reportSlotStatus(InstanceID instanceId, SlotReport slotReport) {
+		checkInit();
+
+		TaskManagerRegistration taskManagerRegistration = taskManagerRegistrations.get(instanceId);
+
+		if (null != taskManagerRegistration) {
+			HashSet<SlotID> slotsToRemove = new HashSet<>(taskManagerRegistration.getSlots());
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: Same as above: If the copy of the collection is just for concurrent iteration, I would make it an `ArrayList`. Builds faster and iterates faster.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/TaskManagerRegistration.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		return slots.remove(slotId);
+	}
+
+	public void addSlot(SlotID slotId) {
+		slots.add(slotId);
+	}
+
+	public void cancelTimeout() {
+		if (null != timeoutFuture) {
+			timeoutFuture.cancel(true);
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/TaskManagerRegistration.java]: I think I would not "interrupt on cancel" here. The work of the future is simple (schedule the call via the main thread executor). An not sure what would actually happen on an interrupt while scheduling the call, but it seems not required.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java [hunk 5]
```diff
@@ -21,519 +21,897 @@
...
+
+				TaskManagerSlot slot = slots.get(slotStatus.getSlotID());
+
+				idle &= slot.isFree();
+			}
+
+			// remove the slots for which we haven't received a slot status message
+			removeSlots(slotsToRemove);
+
+			if (idle) {
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: I assume this breaks the TaskManager timeouts. Every time a heartbeat comes (with a slot report), the TaskManager is detected to be idle, and the timeout is scheduled, overriding the previous timeout (pushing the timeout further back).

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java [hunk 6]
```diff
@@ -21,519 +21,897 @@
...
+	 * given resource profile defines the available resources for the slot. The task manager
+	 * connection can be used to communicate with the task manager.
+	 *
+	 * @param instanceId identifying the task manager on which the slot lives
+	 * @param slotId identifying the slot on the task manager
+	 * @param allocationId which is currently deployed in the slot
+	 * @param resourceProfile of the slot
+	 * @param taskManagerConnection to communicate with the remote task manager
+	 */
+	private void registerSlot(
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: I would pass the `TaskManagerRegistration` into this method. Looking it up again is redundant work and leads to the "what if it is not found" case (where the question comes up whether to log or to throw an exception), which can be avoided.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java [hunk 7]
```diff
@@ -21,519 +21,897 @@
...
+	private void internalRequestSlot(PendingSlotRequest pendingSlotRequest) throws ResourceManagerException {
+		TaskManagerSlot taskManagerSlot = findMatchingSlot(pendingSlotRequest.getResourceProfile());
 
-		if (chosenRequest != null) {
-			final AllocationID allocationId = chosenRequest.getAllocationId();
-			final SlotRequest slotRequest = pendingSlotRequests.remove(allocationId);
+		if (taskManagerSlot != null) {
+			allocateSlot(taskManagerSlot, pendingSlotRequest);
+		} else {
+			final UUID timeoutIdentifier = UUID.randomUUID();
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: A simple (atomically) incrementing `long` is as efficient and cheaper compared to a cryptographically random uid.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java [hunk 8]
```diff
@@ -21,519 +21,897 @@
...
+		// RPC call to the task manager
+		Future<Acknowledge> requestFuture = gateway.requestSlot(
+			slotId,
+			pendingSlotRequest.getJobId(),
+			allocationId,
+			pendingSlotRequest.getTargetAddress(),
+			leaderId,
+			taskManagerRequestTimeout);
+
+		requestFuture.handle(new BiFunction<Acknowledge, Throwable, Void>() {
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: This `BiFunction` seems to be an *identity* function. Can we just skip the `completableFuture` and define the proper handler directly on the `requestFuture`?

**Comment by StephanEwen** [comment]
I have some suggested edits on top of this (not strictly tied to the changes here) in that commit: https://github.com/StephanEwen/incubator-flink/commit/910de21972992d0ed80f232f3a64bf107c6819f2

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/TaskManagerSlot.java [hunk 9]
```diff
@@ -18,15 +18,17 @@
...
 
-import org.apache.flink.runtime.resourcemanager.registration.TaskExecutorRegistration;
+import org.apache.flink.runtime.instance.InstanceID;
+import org.apache.flink.runtime.resourcemanager.registration.TaskExecutorConnection;
+import org.apache.flink.runtime.resourcemanager.slotmanager.PendingSlotRequest;
 
 import static org.apache.flink.util.Preconditions.checkNotNull;
 
 /**
  * A ResourceSlot represents a slot located in TaskManager from ResourceManager's view. It has a unique
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/TaskManagerSlot.java]: +1

File: flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotProtocolTest.java [hunk 10]
```diff
@@ -68,115 +52,79 @@
...
+		scheduledExecutorService.shutdown();
+
+		if (!scheduledExecutorService.isTerminated()) {
+			List<Runnable> runnables = scheduledExecutorService.shutdownNow();
 
-	@Before
-	public void beforeTest(){
-		testRpcService.clearGateways();
+			for (Runnable runnable : runnables) {
+				runnable.run();
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotProtocolTest.java]: You're right. That way it does not make much sense. I think I wanted to call `scheduledExecutorService.awaitTermination` in the if condition in order to wait for the scheduled tasks to be executed.

Will replace it with `Executors.gracefulShutdown`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java (hunk 1 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: Some test cases do. Admittedly, this is not a strong argument, but on the other hand, it does not really hurt either.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java (hunk 2 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: True. Good point.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java (hunk 3 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: Will also fix it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/TaskManagerRegistration.java (hunk 4 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/TaskManagerRegistration.java]: True, will fix this.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java (hunk 8 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: The reason why we have this indirection is because we want to be able to cancel pending slot requests (e.g. if the SlotManager is suspended) from the `SlotManager` side.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java (hunk 6 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: True indeed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java (hunk 5 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: Yes you're right. The timeout logic is faulty. I guess it's better to follow your proposal wrt to the timeout logic (having a periodic task checking all task manager and pending slot requests). That way we wouldn't have to create these timeout identifiers, too.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java (hunk 7 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java]: True, indeed. I think we don't need these identifier anymore if we have a single timeout task which periodically checks all task managers and pending slot requests.

**Comment by tillrohrmann** [comment]
Thanks for the thorough review @StephanEwen. I'll will pull in your commit with the improvements.

Concerning the other points you've raised:

- I think you're right. The timeouts should be configurable. I would like to do as a follow up because the PR is already big enough.

- You're also right that the timeout logic is not properly working. I agree that the (2) approach is probably less brittle. Additionally we would get rid of the timeout identifier generation.

- I think it is fair to assume that the number of slots don't change over the lifetime of a `TaskManager`. I will change the implementation accordingly.

**Comment by tillrohrmann** [comment]
Incorporated your feedback @StephanEwen and rebased onto the latest master.

File: flink-runtime/src/test/resources/log4j-test.properties [hunk 11]
```diff
@@ -16,7 +16,7 @@
 # limitations under the License.
 ################################################################################
 
-log4j.rootLogger=OFF, console
+log4j.rootLogger=DEBUG, console
```
Comment (KurtYoung) [inline flink-runtime/src/test/resources/log4j-test.properties]: Is this change intended or accidental?

Comment (StephanEwen) [inline flink-runtime/src/test/resources/log4j-test.properties]: That must have been accidental and should be reverted...

Comment (tillrohrmann) [inline flink-runtime/src/test/resources/log4j-test.properties]: Thanks for pointing it out @KurtYoung. Will fix it.
